### ISSUE TITLE
[Paper Note] Transformers are RNNs: Fast Autoregressive Transformers with Linear  Attention, Angelos Katharopoulos+, ICML'20

### ISSUE THREAD

**Issue by AkihikoWatanabe** [AkihikoWatanabe/paper_notes#2356]
# URL
- https://arxiv.org/pdf/2006.16236
# Authors
  - Angelos Katharopoulos
  - Apoorv Vyas
  - Nikolaos Pappas
  - François Fleuret
# Abstract
  - Transformers achieve remarkable performance in several tasks but due to their quadratic complexity, with respect to the input's length, they are prohibitively slow for very long sequences. To address this limitation, we express the self-attention as a linear dot-product of kernel feature maps and make use of the associativity property of matrix products to reduce the complexity from $\mathcal{O}\left(N^2\right)$ to $\mathcal{O}\left(N\right)$, where $N$ is the sequence length. We show that this formulation permits an iterative implementation that dramatically accelerates autoregressive transformers and reveals their relationship to recurrent neural networks. Our linear transformers achieve similar performance to vanilla transformers and they are up to 4000x faster on autoregressive prediction of very long sequences.
# Translation (by gpt-4o-mini)
- Transformersは、いくつかのタスクで顕著なパフォーマンスを達成していますが、入力の長さに対して二次的な複雑性を持つため、非常に長いシーケンスに対しては非常に遅くなってしまいます。この制限に対処するために、自己注意をカーネル特徴マップの線形ドット積として表現し、行列積の結合性の特性を利用して、複雑性を$\mathcal{O}\left(N^2\right)$から$\mathcal{O}\left(N\right)$に削減します。ここで、$N$はシーケンスの長さです。この定式化により、自己回帰型Transformersを劇的に加速する反復的な実装が可能になり、再帰型ニューラルネットワークとの関係が明らかになります。我々の線形Transformersは、従来のTransformersと同様のパフォーマンスを達成し、非常に長いシーケンスの自己回帰予測において最大4000倍の速度向上を実現します。
# Summary (by gpt-4o-mini)
- 自己注意をカーネル特徴マップの線形ドット積として表現することで、Transformersの複雑性を$\mathcal{O}\left(N^2\right)$から$\mathcal{O}\left(N\right)$に削減。これにより、自己回帰型Transformersの速度が最大4000倍向上し、従来のパフォーマンスを維持。

**Comment by AkihikoWatanabe** [comment]
関連:
- #1210

**Comment by AkihikoWatanabe** [comment]
pj page:https://linear-transformers.com

**Comment by AkihikoWatanabe** [comment]
Linear Attention（Linear Transformer)を提案した研究。Softmaxが利用されるFull Attentionのsimilarity部分をfeature map φを持つカーネルk(x, y)で一般化し、(3)--(6)の流れでfeature map φ同士は内積が使える性質と、行列積の結合法則を用いて式変換する。
式変換によって、従来のSoftmax Attentionでは、全てのトークンNに対してQ_i * K_j * V_jのメモリが必要だったものを（O(N^2))、各Queryごとに利用される、KVを用いた情報が共通化され（現在のqueryまでのφ(K)とφ(K)V^Tを逐次的に加算して保持するだけで良い）、再利用が可能となりメモリに関するオーダーがO(N)となる。

これらの話は実際にどのような カーネルを利用するかを無視した話だが、たとえばsoftmaxを表現するために必要な指数カーネルの場合はfeature mapが無限次元となるため有限次元のベクトルでは表現できず、厳密に線形化することができない。このため、有限次元かつ厳密なfeature mapを持つ多公式カーネルを用いて近似する。この場合、計算量のオーダーがO(ND^2M)となり（Nがsequence length, Dがqueryとkeyの次元数, Mがvalueの次元数）、softmax attentionの計算量がO(N^2 max(D,M))であることと比較すると、N > D^2である場合に計算効率が改善される。

学習時は並列に学習が可能で、推論時はφ(K_j)V_j^Tを内部状態とみなすと、各タイムステップでRNNのように状態を更新して保持するだけで良い。

（続きはさらに後で読む）

**Comment by AkihikoWatanabe** [comment]
次: DeltaNet
- #5294
- #5295

次の次: Gated DeltaNet
- #4396

次の次の次: Kimi Delta Attention (KDA)
- #3526
